### PR TITLE
feat: custom eslint parser for Vue Vine

### PR DESCRIPTION
**Pull request by ShenQingchuan**
## Why this?

We can't reuse [vue-eslint-parser](https://github.com/vuejs/vue-eslint-parser) because it's made for scanning and analyzing the `.vue` SFC file, and may not satisfies our requirements that well.

We must make our own custom ESTree to represent a VCF(Vine component function).

There'll be a representation of Vue's template in VCF's return statement.

#### Milestones

- [x] Tokenize all items from vine tagged template string node, and adjust position based on global offset.
- [x] Build `VTemplateRoot` ESTree root node.
- [x] analyze used in template variables

## May put in another PR
- [ ] Create `eslint-plugin-vue-vine` package to maintain rules for Vue Vine.

### CODE REVIEW

**Comment by so1ve** [comment]
@ShenQingchuan I see there are many rollup boilerplate code in our repo. Shall we refactor them to make them reuseable, or use [pkgroll](https://github.com/privatenumber/pkgroll) instead? I use it to build the create-vine-app cli in #49

**Comment by ShenQingchuan** [comment]
> @ShenQingchuan I see there are many rollup boilerplate code in our repo. Shall we refactor them to make them reuseable, or use [pkgroll](https://github.com/privatenumber/pkgroll) instead? I use it to build the create-vine-app cli in #49

@so1ve For sure! Any refactor that brings efficient is definitely worth it. 

If we can't avoid do that for 2 times due to there're 2 PRs left here right now, we could just setup them in the same way, and decrease the difference as much as possible so that there'll not be so many conflicts to resolve.

If you chose to do that in one of these 2 PRs, you can go for it and rebase it in the other PR.

Thanks! 💖 You're my GOAT !

**Comment by so1ve** [comment]
I wonder if doing such refactor in these two PRs will cause huge merge conflict... Let's merge these two PRs anyway, thus we can refactor in `main`.

**Comment by ShenQingchuan** [comment]
## Preview video

https://github.com/vue-vine/vue-vine/assets/46062972/d188bc03-3e2f-4327-85d6-1eafe855883c

## Drawbacks now

Based on what's been tested so far, there are some standard rules that still need to be handled specifically for the Vue Vine scenario. For example, StartTag for an HTML tag that takes up multiple lines and has different Attribute rows can cause eslint to report no-multi-spaces. Eslint-plugin-vue redefines a rule with the same name to cover them. Which means we may need to do the same.

**Comment by baiwusanyu-c** [comment]
image

**Comment by so1ve** [comment]
image

**Comment by DaZuiZui** [comment]
image

**Comment by DaZuiZui** [comment]
284512538-7546a8ba-9974-4e08-acc1-0603cb0279b4

File: packages/eslint-parser/src/common/espree.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+/**
+ * Load `espree` from the loaded ESLint.
+ * If the loaded ESLint was not found, just returns `require("espree")`.
+ */
+export function getEspreeFromLinter(): Espree {
+  if (!espreeCache) {
+    espreeCache = getLinterRequire()?.('espree')
+    if (!espreeCache) {
```
Comment (Simon-He95) [inline packages/eslint-parser/src/common/espree.ts:65]: refatctor:
- espreeCache = getLinterRequire()?.('espree') || dependencyEspree

Comment (ShenQingchuan) [reply packages/eslint-parser/src/common/espree.ts:65]: Simon you could use this `Add a suggestion` button to create a review suggestion diff block, and we can apply that diff changes directly on Github web pages. 🙏

File: packages/eslint-parser/src/common/espree.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      espreeCache = dependencyEspree
```
Comment (Simon-He95) [inline packages/eslint-parser/src/common/espree.ts:66]:
```suggestion
    espreeCache = getLinterRequire()?.('espree') || dependencyEspree
```

File: packages/eslint-parser/src/common/eslint-scope.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Load the newest `eslint-scope` from the loaded ESLint or dependency.
+ */
+export function getEslintScope(): typeof escope & {
+  version: string
+} {
+  if (!escopeCache) {
+    escopeCache = getLinterRequire()?.('eslint-scope')
+    if (
+      !escopeCache
```
Comment (Simon-He95) [inline packages/eslint-parser/src/common/eslint-scope.ts:16]:
```suggestion
      !escopeCache // This judgment should not be needed anymore
```

File: packages/eslint-parser/src/common/espree.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  }
+  return linterEspree
+}
+
+/**
+ * Load `espree` from the user dir.
+ */
+export function getEspreeFromUser(): Espree {
+  try {
+    const cwd = process.cwd()
```
Comment (Simon-He95) [inline packages/eslint-parser/src/common/espree.ts:49]:
```suggestion
    const cwd = process.cwd() // Can this be put outside the function?
```

**Pull request merged**